### ISSUE TITLE
Solver RK4 still surprising behavior

### ISSUE THREAD

**Issue by DaluS** [DaluS/GEMMES#113]
I had trouble with the homemade rk4, and decided to verify if our solvers were working well. In consequence, I coded a very simple couple ODE (dampened harmonic oscillator) and compared results with its analytical solution. 

The Homemade RK4 still have a weird behavior. Here is a difference between the numeric and theoric value : 

![image](https://user-images.githubusercontent.com/11523050/142274450-fab6e19b-f9a1-4fe8-80bf-4aae8921ad3c.png)

The slope is 1/2, it should be 4, this system surprisingly behave as a system of order 1/2.

![image](https://user-images.githubusercontent.com/11523050/142275146-117a1c43-7991-4569-9619-1abd88dcbd6b.png)

It seems to me that the code is right, so I do not know why it is not working https://en.wikipedia.org/wiki/Runge%E2%80%93Kutta_methods

```
def _rk4(dydt_func=None, dt=None, y=None, t=None):
    """
    a traditional RK4 scheme, with:
        - y = array of all variables (all ode)
        - dt = fixed time step
    """
    dy1_on_dt = dydt_func(t, y)
    dy2_on_dt = dydt_func(t + dt/2., y + dy1_on_dt * dt/2.)
    dy3_on_dt = dydt_func(t + dt/2., y + dy2_on_dt * dt/2.)
    dy4_on_dt = dydt_func(t + dt, y + dy3_on_dt*dt)
    return (dy1_on_dt + 2*dy2_on_dt + 2*dy3_on_dt + dy4_on_dt) * dt/6.
```

**Comment by Didou09** [comment]
Will try to investigate this in a few weeks (mid-December probably) if the problem is still there by then :-)

**Comment by DaluS** [comment]
Yes ! I'll focus on using the rk8 from scipy for the moment :)

**Comment by DaluS** [comment]
I coded a Lorenz system to check the  qualitative system, for the moment both the rk2-scipy and the homemade rk4 are wrong :) 
![image](https://user-images.githubusercontent.com/11523050/142455778-57a96859-84db-4951-870a-6d767c3cb377.png)

**Comment by DaluS** [comment]
I went a bit further on the subject : 
* an Rk1 has been added for the very-simple case 
* Coded a convergence test (see `testConvergence_DampOsc` and both Rk1/Rk4 are not power-law which imply that either the convergence test is wrong or that the problem is somewhere else 
* `testConvergence_DampOsc` is showing that the first slope estimation is wrong, it might give insight on the problem 
 
![image](https://user-images.githubusercontent.com/11523050/146810022-44c3854e-df95-4659-85d2-156d81618bcd.png)

The way I see it : The initial condition on the first derivative is not respected, even if well noted in `get_summary`. 
![image](https://user-images.githubusercontent.com/11523050/146810281-fd82a66a-d103-49cc-8907-1b661bdd5cbc.png)

Maybe there is an error in the order of resolution ?

**Comment by Didou09** [comment]
So I had a look using the `DampOscillator` with `Perfect` preset (because it is easy to see any problem) and it appears to me that the homemade rk4 behaves correctly if the time step is fine enough.

Below I compare:
* `erk4-scipy` (variable time step optimized by scipy)
* `erk4-homemade` for `dt` = 0.01 s, 0.001 s and 0.0001 s

We can see that the homemade output converges towards the scipy result as dt decreases.
I put the code so you can reproduce it ( I reduce `Tmax` to avoid very large computations).

```
In [79]: hub = pgm.Hub(model='DampOscillator', preset='Perfect', verb=False)

In [80]: hub.set_dparam(key='Tmax', value=4, verb=False)

In [81]: hub.run(solver='eRK4-scipy')

In [82]: dax = hub.plot(key='Energy', color='k')

In [83]: hub.run(solver='eRK4-homemade')

In [84]: dax = hub.plot(key='Energy', color='r', dax=dax)

In [85]: hub.set_dparam(key='dt', value=0.001, verb=False)

In [86]: hub.run(solver='eRK4-homemade')

In [87]: dax = hub.plot(key='Energy', color='b', dax=dax)

In [88]: hub.set_dparam(key='dt', value=0.0001, verb=False)

In [89]: hub.run(solver='eRK4-homemade')

In [90]: dax = hub.plot(key='Energy', color='g', dax=dax)
```

![image](https://user-images.githubusercontent.com/5929562/148283612-9608aea0-3ebf-4bf3-adf8-e021c9d883fd.png)

This is an aspect that we had already discussed [here](https://github.com/DaluS/GEMMES/pull/95#issue-975597009), in PR #95
On my opinion, it shows that we should:
* Recommend `eRK4-scipy` by default
* Learn how to optimize the time steps size to improve `eRK4-homemade`

**Comment by Didou09** [comment]
I am still not sure why eRK8-scipy gives wrong results though

**Comment by Didou09** [comment]
Solved, see PR #130

**Comment by Didou09** [comment]
I m closing this issue as it was solved
